### PR TITLE
add channel protocol handshake

### PR DESCRIPTION
**Pull request by bxq2011hust**

### CODE REVIEW

File: libchannelserver/ChannelSession.h [hunk 1]
```diff
@@ -44,6 +44,12 @@ namespace dev
 {
 namespace channel
 {
+enum class ProtocalVersion : uint32_t
```
Comment (ywy2090) [inline libchannelserver/ChannelSession.h]: wrong word , protocal should be protocol

File: librpc/Rpc.cpp [hunk 2]
```diff
@@ -1007,10 +1007,12 @@ std::string Rpc::sendRawTransaction(int _groupID, const std::string& _rlp)
...
             auto transactionCallback = *currentTransactionCallback;
-            tx.setRpcCallback(
-                [transactionCallback](LocalisedTransactionReceipt::Ptr receipt, bytes input) {
-                    Json::Value response;
-
+            auto clientProtocolversion = (*m_transactionCallbackVersion)();
+            tx.setRpcCallback([transactionCallback, clientProtocolversion](
+                                  LocalisedTransactionReceipt::Ptr receipt, bytes input) {
+                Json::Value response;
+                if (clientProtocolversion > 0)
```
Comment (wheatli) [inline librpc/Rpc.cpp]: what's the meaning of 0 here?

File: libchannelserver/ChannelRPCServer.cpp [hunk 3]
```diff
@@ -221,6 +221,47 @@ void ChannelRPCServer::onDisconnect(
...
 
+void dev::ChannelRPCServer::blockNotify(int16_t _groupID, int64_t _blockNumber)
+{
+    std::string topic = "_block_notify_" + std::to_string(_groupID);
+    std::shared_ptr<dev::channel::TopicChannelMessage> message =
+        std::make_shared<dev::channel::TopicChannelMessage>();
+    message->setType(0x1001);
+    message->setSeq(std::string(32, '0'));
+    message->setResult(0);
+    std::vector<dev::channel::ChannelSession::Ptr> activedSessions = getSessionByTopic(topic);
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libchannelserver/ChannelRPCServer.cpp]: good idea

File: libchannelserver/ChannelRPCServer.cpp [hunk 4]
```diff
@@ -320,12 +359,72 @@ void dev::ChannelRPCServer::onClientEthereumRequest(
...
+    catch (std::exception& e)
+    {
+        CHANNEL_LOG(ERROR) << "onClientHandshake" << boost::diagnostic_information(e);
+    }
+}
+
+void dev::ChannelRPCServer::onClientHeartbeat(
+    dev::channel::ChannelSession::Ptr session, dev::channel::Message::Ptr message)
+{
+    std::string data((char*)message->data(), message->dataSize());
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libchannelserver/ChannelRPCServer.cpp]: @ywy2090 agree hahaha

File: libchannelserver/ChannelSession.h [hunk 5]
```diff
@@ -203,7 +217,9 @@ class ChannelSession : public std::enable_shared_from_this<ChannelSession>
     std::shared_ptr<std::set<std::string> > m_topics;
     ThreadPool::Ptr m_requestThreadPool;
     ThreadPool::Ptr m_responseThreadPool;
-
+    ProtocalVersion m_channelProtocal = ProtocalVersion::v1;
+    ProtocalVersion m_latestchannelProtocal = ProtocalVersion::v2;
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libchannelserver/ChannelSession.h]: I will correct misspell. If we upgrade, then we need to modify m_latestchannelProtocol's value.

File: libinitializer/RPCInitializer.cpp [hunk 6]
```diff
@@ -132,20 +132,9 @@ void RPCInitializer::initConfig(boost::property_tree::ptree const& _pt)
             auto handler = blockChain->onReady([groupID, channelRPCServer](int64_t number) {
                 LOG(INFO) << "Push block notify: " << std::to_string(groupID) << "-" << number;
                 auto c = channelRPCServer.lock();
-
                 if (c)
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libinitializer/RPCInitializer.cpp]: ok, i will modify it.

File: libchannelserver/ChannelRPCServer.cpp [hunk 7]
```diff
@@ -320,12 +359,72 @@ void dev::ChannelRPCServer::onClientEthereumRequest(
...
+    }
+}
+
+void dev::ChannelRPCServer::onClientHandshake(
+    dev::channel::ChannelSession::Ptr session, dev::channel::Message::Ptr message)
+{
+    try
+    {
+        std::string data(message->data(), message->data() + message->dataSize());
+        std::stringstream ss;
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libchannelserver/ChannelRPCServer.cpp]: a handshake only run once with one client, so I don't want to spend time on its performance

File: libchannelserver/ChannelRPCServer.cpp [hunk 8]
```diff
@@ -320,12 +359,72 @@ void dev::ChannelRPCServer::onClientEthereumRequest(
...
+
+        Json::Value value;
+        ss >> value;
+        auto protocolVersion =
+            static_cast<ProtocalVersion>(value.get("HighestSupported", 1).asInt());
+        session->setProtocolVersion(protocolVersion > session->latestProtocolVersion() ?
+                                        session->latestProtocolVersion() :
+                                        protocolVersion);
+
+        auto clientType = value.get("clientType", "Unknow Client Type").asString();
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libchannelserver/ChannelRPCServer.cpp]: ok
my falut

File: libchannelserver/ChannelRPCServer.cpp [hunk 9]
```diff
@@ -221,6 +221,47 @@ void ChannelRPCServer::onDisconnect(
...
+    {
+        message->clearData();
+        if (session->protocolVersion() == ProtocalVersion::v1)
+        {
+            message->setTopicData(topic, (const byte*)content.data(), content.size());
+        }
+        else if (session->protocolVersion() == ProtocalVersion::v2)
+        {
+            message->setTopicData(topic, (const byte*)resp.data(), resp.size());
+        }
```
[the remark this answers is not in the document]
Comment (bxq2011hust) [reply libchannelserver/ChannelRPCServer.cpp]: done

File: librpc/Rpc.cpp (hunk 2 above)
Comment (bxq2011hust) [reply librpc/Rpc.cpp]: Please read the comment. The if is useless for now

**Pull request merged**